### PR TITLE
Support k8s v1.13 on Alicloud

### PR DESCRIPTION
**Pull request by jia-jerry**
**What this PR does / why we need it**:
K8s 1.13.X only support csi-v1.0, so the PR add supports for csi-v1.0
 
**Release note**:
Alicloud csi provider csi-v1.0 is not compatible with csi-0.3.0. So it is suggested we only supports k8s v1.13 on Alicloud

```noteworthy user
Alicloud does now support 1.13 shoot clusters. The support for 1.11 will be dropped. Please refrain from creating 1.11 clusters and recreate your existing clusters with Kubernetes version 1.13.
```

### CODE REVIEW

File: charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml [hunk 1]
```diff
@@ -1,4 +1,68 @@
...
+                fieldRef:
+                  fieldPath: metadata.namespace
+            - name: ADDRESS
+              value: /var/lib/kubelet/plugins/diskplugin.csi.alibabacloud.com/csi.sock
+          imagePullPolicy: "IfNotPresent"
+          volumeMounts:
+            - name: socket-dir
+              mountPath: /var/lib/kubelet/plugins/diskplugin.csi.alibabacloud.com
+      volumes:
+        - name: socket-dir
```
Comment (timuthy) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml:55]: Both deployments are really similar. So, instead you could reuse common parts and use `if` `else` only for the `volume` part.

File: charts/seed-controlplane/charts/kube-controller-manager/templates/kube-controller-manager.yaml [hunk 2]
```diff
@@ -70,8 +70,8 @@ spec:
         {{- end }}
         {{- if not .Values.enableCSI}}
         - --external-cloud-volume-plugin={{ .Values.cloudProvider }}
-        {{- end }}
         - --cloud-config=/etc/kubernetes/cloudprovider/cloudprovider.conf
+        {{- end }}
```
Comment (timuthy) [inline charts/seed-controlplane/charts/kube-controller-manager/templates/kube-controller-manager.yaml:74]: Can you explain why this change is necessary, i.e. why `--cloud-config` is not considered anymore when `.Values.enableCSI == true`?

File: charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml [hunk 3]
```diff
@@ -1,4 +1,68 @@
 {{- if .Values.enabled }}
+{{- if semverCompare ">= 1.13" .Values.kubernetesVersion -}}
```
Comment (rfranzke) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml:2]: Why do we need this `if`-`else` block around the whole deployment?!

File: charts/shoot-core/charts/csi-alicloud/templates/csi-diskplugin-ds.yaml [hunk 4]
```diff
@@ -1,4 +1,108 @@
 {{- if .Values.enabled }}
+{{- if semverCompare ">= 1.13" .Values.kubernetesVersion -}}
```
Comment (rfranzke) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-diskplugin-ds.yaml:2]: Why do we need this `if`-`else` block around the whole daemonset?!

File: charts/shoot-core/charts/csi-alicloud/templates/csi-provisioner-ss.yaml [hunk 5]
```diff
@@ -1,4 +1,59 @@
 {{- if .Values.enabled }}
+{{- if semverCompare ">= 1.13" .Values.kubernetesVersion -}}
```
Comment (rfranzke) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-provisioner-ss.yaml:2]: Why do we need this `if`-`else` block around the whole statefulset?!

File: docs/usage/supported_k8s_versions.md [hunk 6]
```diff
@@ -22,4 +22,4 @@ They are enabled by default in `1.11`. We allow `1.10` but users must make sure
 | Azure          | 1.10.1+         | 1.11.0+         | 1.12.1+         | 1.13.0+         |
 | GCP            | 1.10.0+         | 1.11.0+         | 1.12.1+         | 1.13.0+         |
 | OpenStack      | 1.10.1+         | 1.11.0+         | 1.12.1+         | 1.13.0+         |
-| Alicloud       | unsupported     | 1.11.0+         | unsupported     | unsupported     |
+| Alicloud       | unsupported     | unsupported     | unsupported     | 1.13.0+         |
```
Comment (rfranzke) [inline docs/usage/supported_k8s_versions.md:25]: Why is 1.11 now unsupported?

File: pkg/operation/common/utils.go [hunk 7]
```diff
@@ -444,6 +444,64 @@ func ShouldObjectBeRemoved(obj metav1.Object, gracePeriod time.Duration) bool {
...
+
+	for k, v := range csiFG {
+		featureGates[k] = v
+	}
+
+	return featureGates, nil
+}
+
+// InjectKubeletCSIFeatureGates adds required feature gates for csi when bootstraping Kubelet based on kubernetes version
+func InjectKubeletCSIFeatureGates(kubeVersion string, featureGates map[string]interface{}) (map[string]interface{}, error) {
```
Comment (rfranzke) [inline pkg/operation/common/utils.go]: exactly the same code like above. Can this be merged to one function?

File: pkg/operation/hybridbotanist/cloud_config.go [hunk 8]
```diff
@@ -158,7 +158,19 @@ func (b *HybridBotanist) generateOriginalConfig() (map[string]interface{}, error
 	if kubeletConfig != nil {
 		originalConfig["kubernetes"].(map[string]interface{})["kubelet"].(map[string]interface{})["featureGates"] = kubeletConfig.FeatureGates
 	}
-
+	if userDataConfig.EnableCSI {
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/cloud_config.go]:
```suggestion
	if b.Shoot.UsesCSI() {
```

File: pkg/operation/hybridbotanist/controlplane.go [hunk 9]
```diff
@@ -349,6 +349,20 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
 	}
 	defaultValues["admissionPlugins"] = admissionPlugins
 
+	if b.ShootCloudBotanist.GenerateCloudConfigUserDataConfig().EnableCSI {
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/controlplane.go]:
```suggestion
	if b.Shoot.UsesCSI() {
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml (hunk 3 above)
Comment (jia-jerry) [reply charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml:2]: For easy maintainance.
Actually, we plan only support k8s v1.13.+ because
1. CSI is more stable.
2. CSI Alicloud plugin v1.0 is not compatible with v0.3.0, because they changed the plugin name.

We  need 2 release cycle to retire v1.11.+.
The PR is for the first cycle and supports v1.11(Only for delete) & v1.13.
In the next release cycle, we will remove v1.11 part -- just remove else part

File: charts/shoot-core/charts/csi-alicloud/templates/csi-diskplugin-ds.yaml (hunk 4 above)
Comment (jia-jerry) [reply charts/shoot-core/charts/csi-alicloud/templates/csi-diskplugin-ds.yaml:2]: same reason as previous comment

File: charts/shoot-core/charts/csi-alicloud/templates/csi-provisioner-ss.yaml (hunk 5 above)
Comment (jia-jerry) [reply charts/shoot-core/charts/csi-alicloud/templates/csi-provisioner-ss.yaml:2]: same reason as previous comment

File: docs/usage/supported_k8s_versions.md (hunk 6 above)
Comment (jia-jerry) [reply docs/usage/supported_k8s_versions.md:25]: Not expect users to create k8s v1.11.+

File: pkg/operation/hybridbotanist/controlplane.go [hunk 10]
```diff
@@ -349,6 +349,20 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
 	}
 	defaultValues["admissionPlugins"] = admissionPlugins
 
+	if b.Shoot.UsesCSI() {
+		var featureGates map[string]interface{}
+		if defaultValues["featureGates"] != nil {
+			featureGates = defaultValues["featureGates"].(map[string]interface{})
+		}
+		newFeatures, err := common.InjectCSIFeatureGates(b.ShootVersion(), featureGates)
```
Comment (timuthy) [inline pkg/operation/hybridbotanist/controlplane.go]: Why not reusing `featureGates`?

File: pkg/operation/hybridbotanist/cloud_config.go [hunk 11]
```diff
@@ -158,7 +158,19 @@ func (b *HybridBotanist) generateOriginalConfig() (map[string]interface{}, error
 	if kubeletConfig != nil {
 		originalConfig["kubernetes"].(map[string]interface{})["kubelet"].(map[string]interface{})["featureGates"] = kubeletConfig.FeatureGates
 	}
-
+	if b.Shoot.UsesCSI() {
+		var featureGates map[string]interface{}
+		if originalConfig["kubernetes"].(map[string]interface{})["kubelet"].(map[string]interface{})["featureGates"] != nil {
+			featureGates = originalConfig["kubernetes"].(map[string]interface{})["kubelet"].(map[string]interface{})["featureGates"].(map[string]interface{})
+		}
+		newFeatures, err := common.InjectCSIFeatureGates(b.ShootVersion(), featureGates)
```
Comment (timuthy) [inline pkg/operation/hybridbotanist/cloud_config.go]: Why not reusing `featureGates`?

**Review by timuthy — CHANGES_REQUESTED** [review]

File: charts/shoot-core/charts/csi-alicloud/templates/crd-csidriver.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+{{- if .Values.enabled }}
+{{- if semverCompare ">= 1.13" .Values.kubernetesVersion -}}
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: null
```
Comment (mvladev) [inline charts/shoot-core/charts/csi-alicloud/templates/crd-csidriver.yaml]: this is not needed

File: charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml [hunk 13]
```diff
@@ -1,4 +1,68 @@
...
+    matchLabels:
+      app: csi-attacher
+  template:
+    metadata:
+      annotations:
+        # This, along with the CriticalAddonsOnly toleration below,
+        # marks the pod as a critical add-on, ensuring it gets
+        # priority scheduling and that its resources are reserved
+        # if it ever gets evicted.
+        scheduler.alpha.kubernetes.io/critical-pod: ''
```
Comment (mvladev) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml]: This annotation is deprecated as of `v1.13`.  `priorityClassName: system-cluster-critical` serves the same purpose.

File: charts/shoot-core/charts/csi-alicloud/templates/csi-diskplugin-ds.yaml [hunk 14]
```diff
@@ -1,4 +1,108 @@
...
+    matchLabels:
+      app: csi-disk-plugin-alicloud
+  template:
+    metadata:
+      annotations:
+        # This, along with the CriticalAddonsOnly toleration below,
+        # marks the pod as a critical add-on, ensuring it gets
+        # priority scheduling and that its resources are reserved
+        # if it ever gets evicted.
+        scheduler.alpha.kubernetes.io/critical-pod: ''
```
Comment (mvladev) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-diskplugin-ds.yaml]: Same as above

File: charts/shoot-core/charts/csi-alicloud/templates/csi-diskplugin-ds.yaml [hunk 15]
```diff
@@ -1,4 +1,108 @@
...
+        checksum/secret-csi-diskplugin-alicloud: {{ include (print $.Template.BasePath "/credential-secret.yaml") . | sha256sum }}
+      labels:
+        app: csi-disk-plugin-alicloud
+        origin: gardener
+        garden.sapcloud.io/role: system-component
+    spec:
+      priorityClassName: system-cluster-critical
```
Comment (mvladev) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-diskplugin-ds.yaml]: Use `system-node-critical` instead.

File: charts/shoot-core/charts/csi-alicloud/templates/csi-provisioner-ss.yaml [hunk 16]
```diff
@@ -1,4 +1,59 @@
...
+    matchLabels:
+      app: csi-provisioner
+  template:
+    metadata:
+      annotations:
+        # This, along with the CriticalAddonsOnly toleration below,
+        # marks the pod as a critical add-on, ensuring it gets
+        # priority scheduling and that its resources are reserved
+        # if it ever gets evicted.
+        scheduler.alpha.kubernetes.io/critical-pod: ''
```
Comment (mvladev) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-provisioner-ss.yaml]: Not needed

File: charts/shoot-core/charts/csi-alicloud/templates/csi-snapshotter-ss.yaml [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    matchLabels:
+      app: csi-snapshotter
+  template:
+    metadata:
+      annotations:
+        # This, along with the CriticalAddonsOnly toleration below,
+        # marks the pod as a critical add-on, ensuring it gets
+        # priority scheduling and that its resources are reserved
+        # if it ever gets evicted.
+        scheduler.alpha.kubernetes.io/critical-pod: ''
```
Comment (mvladev) [inline charts/shoot-core/charts/csi-alicloud/templates/csi-snapshotter-ss.yaml]: Not needed

File: charts/seed-controlplane/charts/kube-controller-manager/templates/kube-controller-manager.yaml (hunk 2 above)
Comment (jia-jerry) [reply charts/seed-controlplane/charts/kube-controller-manager/templates/kube-controller-manager.yaml:74]: First, we use out-of-tree cloud controller manager. We do not need cloud-config. 
So if we don't use CSI,  Do we need cloud-config?
@rfranzke @mvladev Please also confirm it. If cloud-controller-manager do not use cloud-config to do volume related provisioning, cloud-config will never be used, right?

But when we use CSI,  kube-controller-manager doesn't need the knowledge of cloud provider.

File: charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml (hunk 1 above)
Comment (jia-jerry) [reply charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml:55]: Please see my previous comment.  k8s v1.11 on will be retired and it is easy for maintain.

File: charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml (hunk 13 above)
Comment (jia-jerry) [reply charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml]: Thanks for your hint. Will fix it.

File: charts/seed-controlplane/charts/kube-controller-manager/templates/kube-controller-manager.yaml (hunk 2 above)
Comment (rfranzke) [reply charts/seed-controlplane/charts/kube-controller-manager/templates/kube-controller-manager.yaml:74]: I think what Jerry says is correct, but I have only little experience with OOT cloud providers yet.

File: charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml (hunk 3 above)
Comment (rfranzke) [reply charts/shoot-core/charts/csi-alicloud/templates/csi-attacher-dep.yaml:2]: ok

File: docs/usage/supported_k8s_versions.md (hunk 6 above)
Comment (rfranzke) [reply docs/usage/supported_k8s_versions.md:25]: ok

**Review by rfranzke — DISMISSED** [review]
/lgtm
